File: 02_framing_character_stuffing/character_stuffing.py
# CN Lab Experiment 2(i): Character Stuffing
# FLAG marks the beginning/end of a frame.
# ESC is inserted before FLAG or ESC when it appears inside data.

FLAG = "F"
ESC = "E"

data = input("Enter data: ")  # Store the original data entered by the user.
stuffed = ""                  # Store the framed and stuffed data.

for ch in data:               # Check every character in the input.
    if ch == FLAG or ch == ESC:  # Special characters need an ESC before them.
        stuffed += ESC            # Insert the escape character.
    stuffed += ch                  # Add the original character.

frame = FLAG + stuffed + FLAG     # Add FLAG at the beginning and end.
print("Stuffed frame:", frame)

File: 02_framing_character_stuffing/CharacterStuffing.java
// CN Lab Experiment 2(i): Character Stuffing

import java.util.Scanner;

public class CharacterStuffing {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        final char FLAG = 'F'; // FLAG identifies the start and end of a frame.
        final char ESC = 'E';  // ESC protects special characters inside data.

        System.out.print("Enter data: ");
        String data = scanner.nextLine(); // Read the original data.

        StringBuilder stuffed = new StringBuilder(); // Store stuffed data.

        for (char ch : data.toCharArray()) { // Process each character.
            if (ch == FLAG || ch == ESC) {  // Check whether it is special.
                stuffed.append(ESC);         // Insert ESC before a special character.
            }
            stuffed.append(ch);              // Add the original character.
        }

        String frame = FLAG + stuffed.toString() + FLAG; // Add frame boundaries.
        System.out.println("Stuffed frame: " + frame);

        scanner.close();
    }
}
